---
title: "Population genetics"
author: "MEC Lab"
date: '2022-08-26'
output: html_document
---
Next steps (following convo on 12/12/2022)
1. Clean up code that's already here
2. Get code from Blair for Admixture and Structure
3. Circle back to relatedness
  a. Relatedness probably doesn't need to be in this - we can probably put in the separate kinship script.


JDS: For more notes, see RA tab in OneNote
Categories:
Start with a little blurb saying, "This assumes that you did x, y, and z upstream".
Initial data QC (check out strataG tutorial that Lisa sent me)
	- Duplicates
	- See https://github.com/lkomoro/Pop-Genomics/blob/master/strataG_new_user_guide.docx
		- Don't worry about LD right now in terms of running for each population (as suggested in the link above)
		- There are scripts corresponding to the coming soon: section at the end of the strataG word doc above in the MEC lab github, maybe green turtles.

Ordination
	- PCA 
	- DAPC

Differentiation
	- Fst, Gst, etc (maybe via strataG, which can run all)
		function overallTest(Msat.g, stats = "all", nrep = 100, write.output=FALSE)
		also function pairwiseTest(Msat.g, stats = c("fst"), nrep = 50, write.output=FALSE) 
	- Population genetic distance
	- Individual genetic distance (relatedness)
		Any other kinship stuff would probably be in a separate repo
	- Structure, fineRADstructure, etc

Diversity metrics
	- Heterozygosity (mayyyyybe FIS)
	
FastStructure/Admixture
	check in with Blair
	try to run structure via strataG


JDS: Include two sentences about what is the purpose of this script. "In this script, we've included code to run the basics if we have a population genetics SNP set that is downstream of filters." We want to run this for most things, but not everything.


LMK to JDS: might be good to add the code to install the required packages if the users dont have them? And particularly note if any of these need versions that need to be installed from sources other than CRAN (eg github). Since I have a new laptop I'm having to reinstall everything, so I've noted below where things had issues

JDS: Include installation instructions
## Load packages
```{r, warning=FALSE, echo = FALSE}
library(tidyverse)
library(dartR) #this calls SNPrelate as a dependency, which seems to have been taken off CRAN
library(poppr)
library(vcfR)
library(adegenet)
library(hierfstat)
library(pophelper)#not on CRAN
library(gridExtra)
library(ggpubr)
library(strataG)#not on CRAN
library(mapdata)
library(viridis)
#library(poppr)#duplicate, listed above?
library(wesanderson)
library(ape)
library(reshape2)
```


## Read in data and create genind object with populations
Note that the population map should be two columns: sample ID and linked population.
LMK to JDS: (1) here add a sentences explaining what the purpose of this chunk is? (2) it might be helpful if we put up example files for the input to show what they should look like/format? The descriptions below are very helpful.

JDS: Look into "here" package. Set up something that can make it easier to change paths. Sometimes this package breaks; if I figure out why, let Lisa know. Can also do what I did with CKMR scripts (specify an object, then use paste0)

JDS: Check for an external doc with explanations (e.g. HackMD) of why we're doing what we're doing. Did I put one of these together already? If so, link to it here; if not, then write one up - can be external (HackMD) or here.

JDS: Include a toy dataset so people can run that dataset through here as an example/vignette (maybe two populations of snails? or cownose ray POPs?).
```{r, echo = FALSE}
#Read in VCF file
in_vcf <- read.vcfR("./data/CCGOM_raw.vcf", convertNA = T)

#Convert vcf to genlight. Save twice: genlight.orig will be preserved if samples are subset and the genlight object will be replaced; if no subsetting is needed, then the file named genlight will be used throughout the rest of the script. 
genlight.orig = genlight <- vcfR2genlight(in_vcf)

#Read in the population map. This should be a two-column tab-delimited text file where the first column contains the name of each individual sample and the second column contains the population. This can be the same as the population map used in Stacks.
popmap.orig <- read_tsv("./data/popmap.txt", col_names = F) %>% 
  dplyr::rename(Indiv = X1, STRATA = X2)

# Filter the population map for whitelisted individuals i.e. those we will keep for the population genetics analysis.
popmap.filtered <- tibble(Indiv = genlight@ind.names) %>% left_join(popmap.orig, by = "Indiv")
genlight@pop <- as.factor(popmap.filtered$STRATA)
summary(genlight@pop)

# Read in a "regions" file. This file should have at a minimum a column for Population, latitude, and longitude. We assume this is a csv (comma-separate) file.
region <- read.csv("regions.csv", header = T)

#Tibbles will only show a couple sig digits for lat and long, but the full coordinates are stored.
latlon <- as.data.frame(cbind(region$Population,
                            region$lat,
                            region$lon)) %>% 
  dplyr::rename(STRATA = V1, lat = V2, lon = V3) %>% 
    mutate(lat = as.numeric(lat), lon = as.numeric(lon))

#Replicate the coordinates for each sample
all_latlon <- popmap.filtered %>% left_join(latlon, by = "STRATA") %>% 
  dplyr::select(-Indiv)

latlon <- all_latlon[,2:3]
genlight@other$latlong<-latlon

genlight_2 <- gl.compliance.check(genlight)


#-------------------Subset the full set of samples for analysis------------------------
#Only run this section if you want to run the analysis on a subset of individuals included in the filtered VCF file. If you want to run on all the samples in the vcf file that was loaded, then skip this section and move on to "Finish creating genlight and genind objects" a few lines below.

#Create a blacklist of individuals that we don't want to include in the analysis so we can drop them from the genlight object

#To drop blacklisted individuals. Note that this will not run if the blacklist solely includes individuals that were already filtered.

#Read in white list of individuals/locs. This should be a text file with a single column containing the names of the individuals that were retained after filtering.
whitelist.indvs <- read.delim("./Whitelist_indivs.txt", header = F, strip.white = T) %>% pull()
genlight_3 <- gl.keep.ind(genlight_2, ind.list = whitelist.indvs)
genlight_3@ind.names

whitelist.locs <- read.delim("./Whitelist_loci.txt", header = F, strip.white = T) %>%
  dplyr::rename(Chrom = V1, Pos = V2) %>%
  arrange(Chrom, Pos)
whitelist.locs_2 <- str_c(whitelist.locs$Chrom, "_", whitelist.locs$Pos) #ISG: WHYYYYYY
genlight_4 <-gl.keep.loc(genlight_3, loc.list = whitelist.locs_2, verbose=0)

rm(genlight, genlight_2, genlight_3)

#-------------------Finish creating genlight and genind objects--------------------------


genind <- gl2gi(genlight_4) # need genind format for some functions
summary(genlight_4@pop)
summary(genind@pop)
```


JDS: PCA doesn't necessarily need to come first. Maybe do duplicate checks first, maybe diversity metrics (heterozygosity, FsT, etc). Include a better description of what the purpose of PCA is.
Go back to strataG and look at the tutorial (which apparently I got real excited about before). We don't ned to use strataG, but we probably want to follow something similar.

With things like IBD, we want to specify that it depends on the hypothesis. This can be a break in the script (e.g. common thigns above, specific thigns below).

## PCA 
Proportion of variation explained by PCA axes
```{r, echo = FALSE}
pca1 <- glPca(genlight_4,center = T, scale = T, nf = 5)
barplot(100*pca1$eig/sum(pca1$eig), col = heat.colors(50), main="PCA Eigenvalues") # retain first 5 axes, incremental decrease after 2
title(ylab="Percent of variance\nexplained", line = 2)
title(xlab="Eigenvalues", line = 1)

#proportion of explained variance by first three axes
a1<-pca1$eig[1]/sum(pca1$eig) # proportion of variation explained by 1st axis
a2<-pca1$eig[2]/sum(pca1$eig) # proportion of variation explained by 2nd axis 
a3<-pca1$eig[3]/sum(pca1$eig) # proportion of variation explained by 3rd axis
pcvar <- data.frame(Axis = c(1:3), Proportion = c(a1,a2,a3))
pcvar
```


### Generate PCA plots
```{r, fig.width = 6,fig.height = 6, echo = FALSE}
# Extract PC scores to color by location/yr:
# Adapted from https://github.com/grunwaldlab/Population_Genetics_in_R/blob/master/gbs_analysis.Rmd#principal-components-analysis
                                    
pca1.scores <- as.data.frame(pca1$scores)
pca1.scores$pop <- pop(genlight_4)
pca1.scores$reg <- pca1.scores$pop

#Make PCA plots
set.seed(9)
num_pops <- length(levels(factor(popmap.filtered$STRATA)))
pop_cols <- get_palette(k = num_pops, palette = "Set1")

# plot PC 1 and 2
ggscatter(pca1.scores, x = "PC1", y = "PC2", shape = "pop", color = "pop",
          palette = pop_cols, ellipse = T, ellipse.level = 0.95,
          xlab = paste0("PC1 (",round(pcvar[1,2]*100,2),"%)"),
          ylab = paste0("PC2 (",round(pcvar[2,2]*100,2),"%)"))

# #plot  PC 2 and 3: not informative for this dataset
ggscatter(pca1.scores, x = "PC2", y = "PC3", shape = "pop", color = "pop",
          palette = pop_cols, ellipse = T, ellipse.level = 0.95,
          xlab = paste0("PC2 (",round(pcvar[2,2]*100,2),"%)"),
          ylab = paste0("PC3 (",round(pcvar[3,2]*100,2),"%)"))
# ggsave("p12.pdf", width = 7, height=7)
```
## PCA 2 - Remove Outliers 1
Proportion of variation explained by PCA axes
```{r, echo = FALSE}
genlight_5<-gl.drop.ind(genlight_4, ind.list = c("RAD-CCGOM_wA05_KEY074_SWFL","RAD-CCGOM_wG05_KEY066_SWFL", "RAD-CCGOM_wB05_KEY048_SWFL", "RAD-CCGOM_wD06_KEY073_SWFL", "RAD-CCGOM_wE12_KKF345R_NGM", "RAD-CCGOM_wF05_KKF073R_NGM"))
pca2 <- glPca(genlight_5,center = T, scale = T, nf = 5)
barplot(100*pca2$eig/sum(pca2$eig), col = heat.colors(50), main="PCA Eigenvalues") # retain first 5 axes, incremental decrease after 2
title(ylab="Percent of variance\nexplained", line = 2)
title(xlab="Eigenvalues", line = 1)

#proportion of explained variance by first three axes
a1.1<-pca2$eig[1]/sum(pca2$eig) # proportion of variation explained by 1st axis
a2.1<-pca2$eig[2]/sum(pca2$eig) # proportion of variation explained by 2nd axis 
a3.1<-pca2$eig[3]/sum(pca2$eig) # proportion of variation explained by 3rd axis
pcvar2 <- data.frame(Axis = c(1:3), Proportion = c(a1.1,a2.1,a3.1))
pcvar2
```

### Generate New PCA plots 1
```{r, fig.width = 6,fig.height = 6, echo = FALSE}
# Extract PC scores to color by location/yr:
# Adapted from https://github.com/grunwaldlab/Population_Genetics_in_R/blob/master/gbs_analysis.Rmd#principal-components-analysis
                                    
pca2.scores <- as.data.frame(pca2$scores)
pca2.scores$pop <- pop(genlight_5)
pca2.scores$reg <- pca2.scores$pop

#Make PCA plots
set.seed(9)
num_pops <- length(levels(factor(popmap.filtered$STRATA)))
pop_cols <- get_palette(k = num_pops, palette = "Set1")

# plot PC 1 and 2
ggscatter(pca2.scores, x = "PC1", y = "PC2", shape = "pop", color = "pop",
          palette = pop_cols, ellipse = T, ellipse.level = 0.95,
          xlab = paste0("PC1 (",round(pcvar2[1,2]*100,2),"%)"),
          ylab = paste0("PC2 (",round(pcvar2[2,2]*100,2),"%)"))

# #plot  PC 2 and 3: not informative for this dataset
ggscatter(pca2.scores, x = "PC2", y = "PC3", shape = "pop", color = "pop",
          palette = pop_cols, ellipse = T, ellipse.level = 0.95,
          xlab = paste0("PC2 (",round(pcvar2[2,2]*100,2),"%)"),
          ylab = paste0("PC3 (",round(pcvar2[3,2]*100,2),"%)"))
# ggsave("p12.pdf", width = 7, height=7)
genind2 <- gl2gi(genlight_5)
```
## PCA 3 - Remove Outliers 2
Proportion of variation explained by PCA axes
```{r, echo = FALSE}
genlight_6<-gl.drop.ind(genlight_5, ind.list = c( "RAD-CCGOM_wD05_PRCS576C+C_DRTO", "RAD-CCGOM_wG08_PRCS512_DRTO", "RAD-CCGOM_wA06_KEY047_SWFL", "RAD-CCGOM_wB02_KEY433_SWFL"))

#"RAD-CCGOM_wA09_PRCS502_DRTO", "RAD-CCGOM_wA12_PRCS632_DRTO", "RAD-CCGOM_wB01_PRCS500C+C_DRTO", "RAD-CCGOM_wB06_PRCS543_DRTO", "RAD-CCGOM_wH10_PRCS635C+C_DRTO",

pca3 <- glPca(genlight_6,center = T, scale = T, nf = 5)
barplot(100*pca3$eig/sum(pca3$eig), col = heat.colors(50), main="PCA Eigenvalues") # retain first 5 axes, incremental decrease after 2
title(ylab="Percent of variance\nexplained", line = 2)
title(xlab="Eigenvalues", line = 1)

#proportion of explained variance by first three axes
a1.2<-pca3$eig[1]/sum(pca3$eig) # proportion of variation explained by 1st axis
a2.2<-pca3$eig[2]/sum(pca3$eig) # proportion of variation explained by 2nd axis 
a3.2<-pca3$eig[3]/sum(pca3$eig) # proportion of variation explained by 3rd axis
pcvar3 <- data.frame(Axis = c(1:3), Proportion = c(a1.2,a2.2,a3.2))
pcvar3
genind3 <- gl2gi(genlight_6)
```
### Generate New PCA plots 2
```{r, fig.width = 6,fig.height = 6, echo = FALSE}
# Extract PC scores to color by location/yr:
# Adapted from https://github.com/grunwaldlab/Population_Genetics_in_R/blob/master/gbs_analysis.Rmd#principal-components-analysis
                                    
pca3.scores <- as.data.frame(pca3$scores)
pca3.scores$pop <- pop(genlight_6)
pca3.scores$reg <- pca3.scores$pop

#Make PCA plots
set.seed(9)
num_pops <- length(levels(factor(popmap.filtered$STRATA)))
pop_cols <- get_palette(k = num_pops, palette = "Set1")

# plot PC 1 and 2
ggscatter(pca3.scores, x = "PC1", y = "PC2", shape = "pop", color = "pop",
          palette = pop_cols, ellipse = T, ellipse.level = 0.95,
          xlab = paste0("PC1 (",round(pcvar3[1,2]*100,2),"%)"),
          ylab = paste0("PC2 (",round(pcvar3[2,2]*100,2),"%)"))

# #plot  PC 2 and 3: not informative for this dataset
ggscatter(pca3.scores, x = "PC2", y = "PC3", shape = "pop", color = "pop",
          palette = pop_cols, ellipse = T, ellipse.level = 0.95,
          xlab = paste0("PC2 (",round(pcvar3[2,2]*100,2),"%)"),
          ylab = paste0("PC3 (",round(pcvar3[3,2]*100,2),"%)"))
# ggsave("p12.pdf", width = 7, height=7)
```
## PCA 4 - Remove Outliers 3
Proportion of variation explained by PCA axes
```{r, echo = FALSE}
genlight_7<-gl.drop.ind(genlight_6, ind.list = c("RAD-CCGOM_wA11_PRCS481_DRTO", "RAD-CCGOM_wC10_PRCS549_DRTO", "RAD-CCGOM_wD10_PRCS703_DRTO", "RAD-CCGOM_wF12_PRCS706_DRTO", "RAD-CCGOM_wF01_KKF356R_NGM", "RAD-CCGOM_wG06_KKF347R_NGM"))
pca4 <- glPca(genlight_7,center = T, scale = T, nf = 5)
barplot(100*pca4$eig/sum(pca4$eig), col = heat.colors(50), main="PCA Eigenvalues") # retain first 5 axes, incremental decrease after 2
title(ylab="Percent of variance\nexplained", line = 2)
title(xlab="Eigenvalues", line = 1)

#proportion of explained variance by first three axes
a1.3<-pca4$eig[1]/sum(pca4$eig) # proportion of variation explained by 1st axis
a2.3<-pca4$eig[2]/sum(pca4$eig) # proportion of variation explained by 2nd axis 
a3.3<-pca4$eig[3]/sum(pca4$eig) # proportion of variation explained by 3rd axis
pcvar4 <- data.frame(Axis = c(1:3), Proportion = c(a1.3,a2.3,a3.3))
pcvar4
```
### Generate New PCA plots 3
```{r, fig.width = 6,fig.height = 6, echo = FALSE}
# Extract PC scores to color by location/yr:
# Adapted from https://github.com/grunwaldlab/Population_Genetics_in_R/blob/master/gbs_analysis.Rmd#principal-components-analysis
                                    
pca4.scores <- as.data.frame(pca4$scores)
pca4.scores$pop <- pop(genlight_7)
pca4.scores$reg <- pca4.scores$pop

#Make PCA plots
set.seed(9)
num_pops <- length(levels(factor(popmap.filtered$STRATA)))
pop_cols <- get_palette(k = num_pops, palette = "Set1")

# plot PC 1 and 2
ggscatter(pca4.scores, x = "PC1", y = "PC2", shape = "pop", color = "pop",
          palette = pop_cols, ellipse = T, ellipse.level = 0.95,
          xlab = paste0("PC1 (",round(pcvar4[1,2]*100,2),"%)"),
          ylab = paste0("PC2 (",round(pcvar4[2,2]*100,2),"%)"))

# #plot  PC 2 and 3: not informative for this dataset
ggscatter(pca4.scores, x = "PC2", y = "PC3", shape = "pop", color = "pop",
          palette = pop_cols, ellipse = T, ellipse.level = 0.95,
          xlab = paste0("PC2 (",round(pcvar4[2,2]*100,2),"%)"),
          ylab = paste0("PC3 (",round(pcvar4[3,2]*100,2),"%)"))
# ggsave("p12.pdf", width = 7, height=7)
```

## Genetic Diversity and FST
Recommend running the pair-wise FST estimate ONCE, then writing a file to read in for next steps to avoid ~20 min run times each time the chunk is run.
```{r, echo = FALSE}
#Calculate pairwise FST
pop.fst <- genet.dist(genind, method = "WC84") # Run time ~20 mins
pop.fst.df <- as.data.frame(as.matrix(round(pop.fst, 3)))
pop.fst.df #Look at matrix of Fst values
write.csv(pop.fst.df,"fst_df.csv")
#pop.fst.df<-read.csv(file="fileName.csv", header = T)

#Format Fst dataframe
pop.fst.tri <- pop.fst.df
pop.fst.tri[lower.tri(pop.fst.df, diag=TRUE)] <- NA
fst.mat = data.matrix(pop.fst.tri)
melted <- melt(fst.mat, na.rm =TRUE)

#Plot heatmap of Fst among populations
ggplot(data = melted, aes(Var2, Var1, fill = value))+ geom_tile(color = "white")+ scale_fill_gradient(low = "white", high = "navy", name="FST")  + ggtitle(expression(atop("Pairwise FST, WC (1984)", atop(italic("N = 113, L = 4200"), ""))))+labs( x = "Sampling Site", y = "Sampling Site") + theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 11, hjust = 1),axis.text.y = element_text(size = 12)) + coord_fixed()
```

## Isolation-by-distance model
```{r, echo=FALSE, warning=FALSE, message=FALSE}
gl.ibd(genlight_4)
```


## Heterozygosity
```{r}
het<-as.data.frame(t(genind@tab))
hetl<-list()
for(q in 1:ncol(het)){
  df<-as.data.frame(het[,q])
  sample<-colnames(het)[q]
  df<-as.data.frame(df[!is.na(df),])
  df2<-as.data.frame(df[df$`df[!is.na(df), ]` == 1,])
  nHet<-nrow(df2)
  nTot<-nrow(df)
  hTot<-nHet/nTot
  hdf<-as.data.frame(cbind(sample, hTot))
  hdf$pop<-str_sub(hdf$sample, -4, -1)
  hetl[[q]]<-hdf
}
hetdf<-do.call(rbind, hetl)
hetdf$hTot<-as.numeric(hetdf$hTot)

ggscatter(hetdf, x = "pop", y = "hTot", color = "pop", palette = pop_cols,
          xlab = "Population", ylab = "Heterozygosity", legend = "none") +
  font("xlab", face = "bold") + font("ylab", face = "bold")
```


## DAPC 
JDS: move DAPC up to just below PCA.

### Step 1: Run DAPC without any prior information
The chunk here is to determine the optimal value of K (clusters) to use for downstream analyses using k-means clustering. As per the documenttaion from `find.clusters`:

"The rule of thumb consists in increasing K until it no longer leads to an appreciable improvement of fit (i.e., to a decrease of BIC). In the most simple models (island models), BIC decreases until it reaches the optimal K, and then increases. In these cases, our rule amounts to choosing the lowest K. In other models such as stepping stones, the decrease of BIC often continues after the optimal K, but is much less steep."
```{r, echo=TRUE}
n_individuals <- nrow(popmap.filtered)
n_pops <- length(levels(factor(popmap.filtered$STRATA)))

grp_all <- find.clusters(genind, max.n.clust=n_pops, n.pca=200,
                         choose.n.clust = FALSE)

BIC<-as.data.frame(cbind(seq(1,n_pops,1), grp_all$Kstat))

ggline(BIC, x = "V1", y = "V2", plot_type = "b",
       col = "navy",
       xlab = "Number of clusters (K)",
       ylab = "BIC Value",
       title = "Selection of optimum number of clusters (K)") + font("xlab", face = "bold") + font("ylab", face = "bold")
grp_all$Kstat

```


### Step 2: Based on the output of the plot above, we select the optimal value of K and observe the number of samples per cluster. Set k to the value with the lowest BIC score.
```{r, echo=TRUE}
k <- 4
grp_all <- find.clusters(genind, max.n.clust=n_pops, n.pca=200, n.clust = k)
grp_all$size
```

### Step 3: Run DAPC and generate the plots for visualizing
Selecting the right number of DAs is important as it impacts the downstream outputs, and too many PCs can lead to over-fitting the data.
```{r}
n.da <- 6 #Set number of DAs
dapc <- dapc(genind, grp_all$grp, n.pca=60, n.da = n.da, var.contrib = TRUE)
pal <- get_palette("Accent", k = k)
dpca_result <- scatter(dapc, col=pal, scree.pca = TRUE,
                       pch = 20, cell = 0, cstar = 1,
                       solid = 0.8, cex = 3, clab = 1)
set.seed(4)
#contrib<-loadingplot(dapc$var.contr,axis=2, thres=.002,lab.jitter=1)
compoplot(dapc, posi="topleft", txt.leg=paste("Cluster",c(1:k)),
          ncol=1, xlab="Individuals", col=pal, lab=genind@pop, show.lab = T)

dapc$IND <- row.names(dapc$ind.coord)

dapc_info <- as.data.frame(cbind(dapc$IND, dapc$ind.coord, grp_all$grp))

colnames(dapc_info) <- c("IND","DPC1", "DPC2", "DPC3", "K")
colnames(dapc_info) <- c("IND", c(paste0("DPC", 1:(n.da-3))), "K")
dapc_info$SITE <- str_sub(dapc_info$IND, -4,-1) #Takes first two letters of the Indv name and adds to the columns "SITE"
dapc_info
```

### Step 4: Now use results from the first DAPC to inform a secondary DAPC

**JDS Q for Blair:** Why are we calling this a prior? Aren't we just empirically determining the best number of PCs to retain for DAPC?
```{r, echo=TRUE}
set.seed(5); dapc_a_score <- dapc(genind, n.pca=100, n.da=n.da) 
temp_score <- optim.a.score(dapc_a_score) #Determine best number of PCs to retain for DAPC analysis
n.pc <- temp_score$best #Save best number of PCs
dapc2 <-dapc(genind,genind@pop, n.pca = n.pc, n.da = n.da)
dapc2$IND <- row.names(dapc2$ind.coord)
dapc2_info <- as.data.frame(cbind(dapc2$IND, dapc2$ind.coord, grp_all$grp))

dpca_result <- scatter(dapc2, scree.pca = TRUE,
                       pch = 20, cell = 0, cstar = 1,
                       solid = 0.8, cex = 3, clab = 1)

contrib <- loadingplot(dapc2$var.contr, axis=1, thres=.0004, lab.jitter=1)
contrib <- loadingplot(dapc2$var.contr, axis=2, thres=.0004, lab.jitter=1)
#freq <- tab(genind2genpop(genind[loc=c("NW_026041555.1_510841")]),freq=TRUE)


load_dpca2 <- as.data.frame(dapc2$var.contr)
write.table(load_dpca2, "Loadings_SNPs.txt", sep="\t", row.names=FALSE, quote=FALSE)
percent= dapc2$eig/sum(dapc2$eig)*100
barplot(percent, main = "Percent of genetic variance explained by eigenvectors", ylim = c(0, max(percent) + 10), names.arg = paste0("E", c(1:length(percent))), ylab = "Percent", xlab = "Eigenvector")
dapc_prior=as.data.frame(dapc2$ind.coord)
write.table(dapc_prior, "DAPC_prior_results.txt", quote=F, sep="\t", row.names=TRUE)

### Add information to the tab results of the DPCA
dapc_prior$IND <- row.names(dapc_prior)

### Add site info
dapc_prior$SITE <- str_sub(dapc_prior$IND, -4,-1) #Again, assumes the site is the first two characters of the individual names
dapc_prior$SITE <- gsub ('_', '', dapc_prior$SITE)

### Add 'region' info:
dapc_prior <- merge(dapc_prior, region, by.x = "SITE", by.y = "Population")

### Make a ggplot graph representing the DAPC for the first and second axes for the regions
ggscatter(dapc_prior, x = "LD1", y = "LD2", shape = 21, fill = "SITE", size = 3,
          xlab = "DA1", ylab = "DA2", palette = pop_cols)

compoplot(dapc2)
### Save the ggplot graph
ggsave("DAPC_prior.pdf",width=12,height=10,dpi=600,units="cm",useDingbats=F)

```


JDS: we want to add Structure to the script as well. We can potentially call Structure via strataG. Check in with Blair about Structure and whether he has scripts.

## END MAIN SCRIPT
